### PR TITLE
[REEF-1772] Implement IPipelineDataConverter in REEF.NET for arrays

### PR DESCRIPTION
**Pull request by rogancarr**
This issue is addressed by providing a set of concrete implemenations of IPipelineDataConverter for arrays of type:
	float[];
	double[];
	int[]; and
	long[].

Unit tests have been written for these, and abstracted in a way to cover additional array-types with minimal effort.

### CODE REVIEW

File: lang/cs/Org.Apache.REEF.Network/Group/Pipelining/Impl/DoubleArrayPipelineDataConverter.cs [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Utilities.Logging;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.Serialization;
+
+namespace Org.Apache.REEF.Network.Group.Pipelining.Impl
+{
+    public sealed class DoubleArrayPipelineDataConverter : IPipelineDataConverter<double[]>
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Pipelining/Impl/DoubleArrayPipelineDataConverter.cs]: Please provide xml docs for all `public` classes and members.

File: lang/cs/Org.Apache.REEF.Network/Group/Pipelining/Impl/DoubleArrayPipelineDataConverter.cs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+        private DoubleArrayPipelineDataConverter([Parameter(typeof(GroupCommConfigurationOptions.PipelineMessageSize))] int pipelineMessageSize)
+        {
+            if (pipelineMessageSize <= 0)
+            {
+                throw new ArgumentException("PipelinePackageElements must be strictly positive");
+            }
+            _pipelineMessageSize = pipelineMessageSize;
+        }
+
+        public List<PipelineMessage<double[]>> PipelineMessage(double[] message)
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Pipelining/Impl/DoubleArrayPipelineDataConverter.cs]: This makes me question our overall design for the pipelined messages. Returning `List` here is bound to create problems as we can't actually send the messages in a pipelined fashion: The `List` needs to exist first. A much better approach would be to return a `Enumerable` or an `Enumerator`, or even an `Observable`.

File: lang/cs/Org.Apache.REEF.Network/Group/Pipelining/Impl/FloatArrayPipelineDataConverter.cs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+using Org.Apache.REEF.Tang.Annotations;
+using Org.Apache.REEF.Utilities.Logging;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.Serialization;
+
+namespace Org.Apache.REEF.Network.Group.Pipelining.Impl
+{
+    public sealed class FloatArrayPipelineDataConverter : IPipelineDataConverter<float[]>
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Pipelining/Impl/FloatArrayPipelineDataConverter.cs]: Is there a way to capture all of these cases of arrays of primitive types via type parameters?

File: lang/cs/Org.Apache.REEF.Network/Group/Pipelining/Impl/DoubleArrayPipelineDataConverter.cs (hunk 2 above)
Comment (rogancarr) [inline lang/cs/Org.Apache.REEF.Network/Group/Pipelining/Impl/DoubleArrayPipelineDataConverter.cs]: 100% agree here. This is something I have been thinking about too. I think the next step should be to see if we can refactor these all out to use one of the interfaces you suggest. I currently am not sure how embedded `List` is in our current implementation, but I don't think it's going to be too tight, as we don't really use it to its full capacity yet.

Would you be okay putting that in a separate pull request?

File: lang/cs/Org.Apache.REEF.Network/Group/Pipelining/Impl/FloatArrayPipelineDataConverter.cs (hunk 3 above)
Comment (rogancarr) [inline lang/cs/Org.Apache.REEF.Network/Group/Pipelining/Impl/FloatArrayPipelineDataConverter.cs]: Good idea. I think I could do something similar to what I do with the unit tests.

**Comment by rogancarr** [comment]
Thanks for your comments, Markus! I've updated the PR.

**Comment by markusweimer** [comment]
> (note that some of my responses are hidden now in the "outdated" sections.

Yes, that happens when you rebase before pushing. During a code review, it is often easier to just put new commits to the branch under review.

File: lang/cs/Org.Apache.REEF.Network/Group/Pipelining/Impl/DoubleArrayPipelineDataConverter.cs (hunk 2 above)
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network/Group/Pipelining/Impl/DoubleArrayPipelineDataConverter.cs]: > Would you be okay putting that in a separate Issue?

yes. This PR is meaningful progress. So, let's merge it :)

File: lang/cs/Org.Apache.REEF.Network.Tests/GroupCommunication/PipeliningTests.cs [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+using System;
+using System.Collections.Generic;
+using Xunit;
+
+namespace Org.Apache.REEF.Network.Tests.GroupCommunication
+{
+    /// <summary>
+    /// Defines pipelining component tests
+    /// </summary>
+    public class PipeliningTests
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Network.Tests/GroupCommunication/PipeliningTests.cs]: Please add a test for the empty array.

Comment (rogancarr) [inline lang/cs/Org.Apache.REEF.Network.Tests/GroupCommunication/PipeliningTests.cs]: Good find. I added a test for both the empty array and the null array.

Null array passes through and produces a null array.

Empty array passes through and produces an empty array.

**Comment by rogancarr** [comment]
Thanks for the review!

**Comment by markusweimer** [comment]
This is merged. I edited the commit message a bit in the process, have a look [here](https://github.com/apache/reef/commit/affd9735d7cd63c8576b89dc119ef1c1f61b053a)
